File: app/src/main/java/com/qiaoqiao/ds/DsLoadedCallback.java
package com.qiaoqiao.ds;

import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import com.google.api.services.vision.v1.model.BatchAnnotateImagesResponse;
import com.google.api.services.vision.v1.model.Status;
import com.qiaoqiao.backend.model.translate.Data;
import com.qiaoqiao.backend.model.wikipedia.WikiResult;
import com.qiaoqiao.ds.database.HistoryItem;
import com.qiaoqiao.utils.LL;

import java.io.IOException;
import java.util.List;
import java.util.Map;
import java.util.Set;

import io.realm.Realm;


public abstract class DsLoadedCallback {
	public void onSaveHistory(@Nullable  final byte[] byteArray, @Nullable  final Uri imageUri, @NonNull final BatchAnnotateImagesResponse response) {
		final Realm realm = Realm.getDefaultInstance();
		realm.executeTransactionAsync(bgRealm -> {
			HistoryItem historyItem = bgRealm.createObject(HistoryItem.class);
			if (imageUri != null) {
				historyItem.setImageUri(imageUri.toString());
			}
			historyItem.setByteArray(byteArray);
			historyItem.setSavedTime(System.currentTimeMillis());
			try {
				historyItem.setJsonText(response.toPrettyString());
			} catch (IOException e) {
				LL.d("Saved history fail.");
			}
		}, () -> LL.d("Saved history successfully."), error -> LL.d("Saved history fail."));
	}

	public void onVisionResponse(BatchAnnotateImagesResponse response) {
		if (response == null) {
			LL.d("response is NULL.");
			return;
		}
		try {
			LL.d("response is OK.");
			LL.d(response.toPrettyString());
		} catch (IOException e) {
			LL.d("response is available but IO problem: " + e.toString());
		}
	}

	public void onKnowledgeResponse(WikiResult result) {
		LL.d("response of wiki: " + result.toString());
	}

	public void onError(@NonNull Status status) {
		if (!TextUtils.isEmpty(status.getMessage())) {
			LL.e(status.getMessage());
		} else {
			LL.e("Vision API on some error:");
			final List<Map<String, Object>> details = status.getDetails();
			for (Map<String, Object> mapDetail : details) {
				Set<String> mapDetailKeys = mapDetail.keySet();
				for (String key : mapDetailKeys) {
					LL.e(mapDetail.get(key)
					              .toString());
				}
			}
		}
	}

	public void onException(@NonNull Exception e) {
		LL.e(e.toString());
	}

	public void onTranslateData(@NonNull Data translateData) {
		LL.d("size of responses of cloud translate: " + translateData.getTranslations().length);
	}
}

File: app/src/main/java/com/qiaoqiao/repository/DsLoadedCallback.kt
package com.qiaoqiao.repository

import android.net.Uri
import android.text.TextUtils
import com.google.api.services.vision.v1.model.BatchAnnotateImagesResponse
import com.google.api.services.vision.v1.model.Status
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.UploadTask
import com.qiaoqiao.core.product.model.ProductEntity
import com.qiaoqiao.repository.backend.model.translate.Data
import com.qiaoqiao.repository.backend.model.wikipedia.WikiResult
import com.qiaoqiao.repository.backend.model.wikipedia.geo.GeoResult
import com.qiaoqiao.repository.database.HistoryItem
import com.qiaoqiao.utils.LL
import io.realm.Realm
import java.io.IOException

abstract class DsLoadedCallback {
    inline fun pushOnFirebase(bytes: ByteArray, noinline err: (Exception) -> Unit, crossinline succ: (UploadTask.TaskSnapshot) -> Unit) {
        val currentUser = FirebaseAuth.getInstance()
                .currentUser ?: return
        val rootRef = FirebaseStorage.getInstance().reference
        val imageRef = rootRef.child("images/" + currentUser.uid + "/" + System.currentTimeMillis() + ".jpg")
        val uploadTask = imageRef.putBytes(bytes)
        uploadTask.addOnFailureListener { err(it) }.addOnSuccessListener { succ(it) }
    }

    fun saveOnLocalHistory(imageUri: Uri, json: String) {
        val realm = Realm.getDefaultInstance()
        realm.executeTransactionAsync({ bgRealm ->
            val historyItem = bgRealm.createObject(HistoryItem::class.java)
            historyItem.imageUri = imageUri.toString()
            historyItem.jsonText = json
            historyItem.savedTime = System.currentTimeMillis()
        }, { LL.d("Saved history successfully.") }, { LL.d("Saved history fail.") })
    }

    open fun onVisionResponse(response: BatchAnnotateImagesResponse?) {
        if (response == null) {
            LL.d("response is NULL.")
            return
        }
        try {
            LL.d("response is OK.")
            LL.d(response.toPrettyString())
        } catch (e: IOException) {
            LL.d("response is available but IO problem: " + e.toString())
        }

    }

    open fun onKnowledgeResponse(result: WikiResult) {
        LL.d("response of wiki: " + result.toString())
    }

    open fun onGeosearchResponse(result: GeoResult) {
        LL.d("response of wiki: " + result.toString())
    }

    open fun onVisionApiError(status: Status) {
        if (!TextUtils.isEmpty(status.message)) {
            LL.e(status.message)
        } else {
            LL.e("Vision API on some error:")
            val details = status.details
            for (mapDetail in details) {
                val mapDetailKeys = mapDetail.keys
                for (key in mapDetailKeys) {
                    LL.e(mapDetail[key]
                            .toString())
                }
            }
        }
    }

    open fun onKnowledgeResponse(products: List<ProductEntity>) {
    }

    open fun onException(e: Exception) {
        LL.e(e.toString())
    }

    open fun onTranslateData(translateData: Data) {
        LL.d("size of responses of cloud translate: " + translateData.translations.size)
    }

    open fun onImageLoad(imageLocation: Uri) {
        LL.d("loaded an image at location: " + imageLocation.toString())

    }

    open fun onImageLoad(imageBytes: ByteArray) {
        LL.d("loaded an image with size: " + imageBytes.size)
    }

    open fun onSomeThingSuccessfully() {

    }

    open fun onSomeThingUnsuccessfully() {

    }
}